File: BackEnd-Siukat/services/ayah.go
package services

import (
	"BackEnd-Siukat/config"
	"BackEnd-Siukat/models"
	"time"
)

type AyahService struct{}

func (s *AyahService) Add(req models.Ayah, atribut string) (models.Ayah, error) {
	db := config.DB
	req.Atribut = atribut
	err := db.Create(&req).Error
	return req, err
}

func (s *AyahService) Edit(data map[string]interface{}, noPeserta string, atribut string) (models.Ayah, error) {
	db := config.DB
	var ayah models.Ayah

	// SURGICAL FIX: Use Count instead of First to avoid "Scan error" on corrupted existing data
	var count int64
	db.Model(&models.Ayah{}).Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).Count(&count)

	if count == 0 {
		// 1. Jika BELUM ADA, maka Create
		data["no_peserta"] = noPeserta
		data["atribut"] = atribut
		if errCreate := db.Model(&models.Ayah{}).Create(data).Error; errCreate != nil {
			return models.Ayah{}, errCreate
		}
	} else {
		// 2. Jika SUDAH ADA, maka Update menggunakan Table/Where (aman dari Scan error)
		if errUpdate := db.Table("tb_ayah").Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).Updates(data).Error; errUpdate != nil {
			return models.Ayah{}, errUpdate
		}
	}

	// 3. Ambil data terbaru untuk return (Sekarang Scan error harusnya hilang karena data sudah diperbaiki oleh Updates)
	err := db.Preload("Pekerjaan").Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&ayah).Error
	return ayah, err
}

func (s *AyahService) AddLog(user models.Ayah, atribut string, executor string, timestamp *time.Time) error {
	db := config.DB
	logAyah := models.LogAyah{
		NoPeserta:        user.NoPeserta,
		StatusAyah:       user.StatusAyah,
		NamaAyah:         user.NamaAyah,
		NikAyah:          user.NikAyah,
		TeleponAyah:      user.TeleponAyah,
		AlamatAyah:       user.AlamatAyah,
		ProvinsiAyah:     user.ProvinsiAyah,
		KabkotAyah:       user.KabkotAyah,
		KecamatanAyah:    user.KecamatanAyah,
		PekerjaanAyah:    user.PekerjaanAyah,
		PenghasilanAyah:  user.PenghasilanAyah,
		SampinganAyah:    user.SampinganAyah,
		ScanKtpAyah:      user.ScanKtpAyah,
		ScanSlipAyah:     user.ScanSlipAyah,
		TempatLahirAyah:  user.TempatLahirAyah,
		TanggalLahirAyah: user.TanggalLahirAyah,
		Atribut:          atribut,
		Executor:         executor,
		Timestamp:        timestamp,
	}
	return db.Create(&logAyah).Error
}

func (s *AyahService) GetByLoggedIn(noPeserta string) (models.Ayah, error) {
	db := config.DB
	var ayah models.Ayah
	
	// 1. PRIORITAS: Cek data 'sanggah' dulu dengan relasi lengkap
	err := db.Preload("Pekerjaan").Preload("Provinsi").Preload("Kabkot").Preload("Kecamatan").
		Where("no_peserta = ? AND atribut = ?", noPeserta, "sanggah").First(&ayah).Error
	
	if err == nil {
		return ayah, nil
	}

	// 2. FALLBACK: Jika tidak ada sanggah, ambil data 'original' dengan relasi lengkap
	err = db.Preload("Pekerjaan").Preload("Provinsi").Preload("Kabkot").Preload("Kecamatan").
		Where("no_peserta = ? AND atribut = ?", noPeserta, "original").First(&ayah).Error
	
	if err != nil {
		// Jika beneran nggak ada data sama sekali atau Error scanning
		return models.Ayah{NoPeserta: noPeserta}, nil
	}

	return ayah, nil
}

// CheckData — parity dengan ayah.prototype.checkData di Node.js
func (s *AyahService) CheckData(noPeserta string, uktTinggi string) (bool, error) {
	db := config.DB
	var ayah models.Ayah
	if err := db.Where("no_peserta = ?", noPeserta).First(&ayah).Error; err != nil {
		return false, err
	}

	data := map[string]interface{}{
		"no_peserta":        ayah.NoPeserta,
		"status_ayah":       ayah.StatusAyah,
		"nama_ayah":         ayah.NamaAyah,
		"nik_ayah":          ayah.NikAyah,
		"telepon_ayah":      ayah.TeleponAyah,
		"alamat_ayah":       ayah.AlamatAyah,
		"provinsi_ayah":     ayah.ProvinsiAyah,
		"kabkot_ayah":       ayah.KabkotAyah,
		"kecamatan_ayah":    ayah.KecamatanAyah,
		"pekerjaan_ayah":    ayah.PekerjaanAyah,
		"penghasilan_ayah":  ayah.PenghasilanAyah,
		"sampingan_ayah":    ayah.SampinganAyah,
		"scan_ktp_ayah":     ayah.ScanKtpAyah,
		"scan_slip_ayah":    ayah.ScanSlipAyah,
		"tempat_lahir_ayah": ayah.TempatLahirAyah,
		"tanggal_lahir_ayah": ayah.TanggalLahirAyah,
		"atribut":           ayah.Atribut,
	}

	if ayah.StatusAyah == "wafat" {
		return ayah.NamaAyah != "", nil
	}

	if uktTinggi == "ya" {
		delete(data, "nik_ayah")
		delete(data, "pekerjaan_ayah")
		delete(data, "penghasilan_ayah")
		delete(data, "scan_ktp_ayah")
		delete(data, "scan_slip_ayah")
	}

	if ayah.PekerjaanAyah == "11" { // Tidak bekerja
		delete(data, "penghasilan_ayah")
	}

	delete(data, "sampingan_ayah")

	for _, val := range data {
		if val == nil || val == "" || val == 0 {
			return false, nil
		}
	}

	return true, nil
}

File: ZhammDataBeLama/siukat-service-dev_snmptn_2022/services/ayah.js
var Ayah = require("../models/ayah");
var LogAyah = require("../models/log_ayah");
var Provinsi = require("../models/provinsi");
var Kabkot = require("../models/kabkot");
var Kecamatan = require("../models/kecamatan");
var Pekerjaan = require("../models/pekerjaan");

var ayah = function () { };

ayah.prototype.add = function (req, atribut) {
    return new Promise(function (resolve, reject) {
        Ayah.create(
            {
                no_peserta: req.no_peserta,
                status_ayah: req.status_ayah,
                nama_ayah: req.nama_ayah,
                nik_ayah: req.nik_ayah,
                telepon_ayah: req.telepon_ayah,
                alamat_ayah: req.alamat_ayah,
                provinsi_ayah: req.provinsi_ayah,
                kabkot_ayah: req.kabkot_ayah,
                kecamatan_ayah: req.kecamatan_ayah,
                pekerjaan_ayah: req.pekerjaan_ayah,
                penghasilan_ayah: req.penghasilan_ayah,
                sampingan_ayah: req.sampingan_ayah,
                scan_ktp_ayah: req.scan_ktp_ayah,
                scan_slip_ayah: req.scan_slip_ayah,
                tempat_lahir_ayah: req.tempat_lahir_ayah,
                tanggal_lahir_ayah: req.tanggal_lahir_ayah,
                atribut: atribut,
            },
            {
                fields: [
                    "no_peserta",
                    "status_ayah",
                    "nama_ayah",
                    "nik_ayah",
                    "telepon_ayah",
                    "alamat_ayah",
                    "provinsi_ayah",
                    "kabkot_ayah",
                    "kecamatan_ayah",
                    "pekerjaan_ayah",
                    "penghasilan_ayah",
                    "sampingan_ayah",
                    "scan_ktp_ayah",
                    "scan_slip_ayah",
                    "tempat_lahir_ayah",
                    "tanggal_lahir_ayah",
                    "atribut",
                ],
            }
        )
            .then(function (response) {
                resolve(response);
            })
            .catch(function (err) {
                reject(err);
            });
    });
};

ayah.prototype.edit = function (req, no_peserta, atribut) {
    return new Promise(function (resolve, reject) {
        var data = {
            nama_ayah: req.body.nama_ayah,
            status_ayah: req.body.status_ayah,
            nik_ayah: req.body.status_ayah == "wafat" ? "" : req.body.nik_ayah,
            telepon_ayah:
                req.body.status_ayah == "wafat" ? "" : req.body.telepon_ayah,
            alamat_ayah: req.body.status_ayah == "wafat" ? "" : req.body.alamat_ayah,
            provinsi_ayah:
                req.body.status_ayah == "wafat" ? "" : req.body.provinsi_ayah,
            kabkot_ayah: req.body.status_ayah == "wafat" ? "" : req.body.kabkot_ayah,
            kecamatan_ayah:
                req.body.status_ayah == "wafat" ? "" : req.body.kecamatan_ayah,
            pekerjaan_ayah:
                req.body.status_ayah == "wafat" ? "" : req.body.pekerjaan_ayah,
            penghasilan_ayah:
                req.body.status_ayah == "wafat" ? "" : req.body.penghasilan_ayah,
            sampingan_ayah:
                req.body.status_ayah == "wafat" ? "" : req.body.sampingan_ayah,
            tempat_lahir_ayah:
                req.body.status_ayah == "wafat" ? "" : req.body.tempat_lahir_ayah,
            tanggal_lahir_ayah:
                req.body.status_ayah == "wafat" ? "" : req.body.tanggal_lahir_ayah,
            // scan_ktp_ayah: '',
            // scan_slip_ayah: ''
        };

        // Jika tidak bekerja
        // if (req.body.pekerjaan_ayah === "11") {
        //   data.penghasilan_ayah = "";
        //   data.sampingan_ayah = "";
        //   data.scan_slip_ayah = "";
        // }

        if (req.files["file_scan_ktp_ayah"] !== undefined) {
            data.scan_ktp_ayah =
                req.body.status_ayah == "wafat"
                    ? ""
                    : req.files["file_scan_ktp_ayah"][0].filename;
        }

        if (req.files["file_scan_slip_ayah"] !== undefined) {
            data.scan_slip_ayah =
                req.body.status_ayah == "wafat"
                    ? ""
                    : req.files["file_scan_slip_ayah"][0].filename;
        }

        Ayah.update(data, {
            where: {
                no_peserta: no_peserta,
                atribut: atribut,
            },
        })
            .then(function (response) {
                resolve({
                    ...response,
                    no_peserta: no_peserta,
                    atribut: atribut,
                });
            })
            .catch(function (err) {
                reject(err);
            });
    });
};

ayah.prototype.addLog = function (users, atribut, executor, timestamp) {
    return new Promise(function (resolve, reject) {
        LogAyah.create(
            {
                no_peserta: users.no_peserta,
                status_ayah: users.status_ayah,
                nama_ayah: users.nama_ayah,
                nik_ayah: users.nik_ayah,
                telepon_ayah: users.telepon_ayah,
                alamat_ayah: users.alamat_ayah,
                provinsi_ayah: users.provinsi_ayah,
                kabkot_ayah: users.kabkot_ayah,
                kecamatan_ayah: users.kecamatan_ayah,
                pekerjaan_ayah: users.pekerjaan_ayah,
                penghasilan_ayah: users.penghasilan_ayah,
                sampingan_ayah: users.sampingan_ayah,
                scan_ktp_ayah: users.scan_ktp_ayah,
                scan_slip_ayah: users.scan_slip_ayah,
                tempat_lahir_ayah: users.tempat_lahir_ayah,
                tanggal_lahir_ayah: users.tanggal_lahir_ayah,
                atribut: atribut,
                executor: executor,
                timestamp: timestamp,
            },
            {
                fields: [
                    "no_peserta",
                    "status_ayah",
                    "nama_ayah",
                    "nik_ayah",
                    "telepon_ayah",
                    "alamat_ayah",
                    "provinsi_ayah",
                    "kabkot_ayah",
                    "kecamatan_ayah",
                    "pekerjaan_ayah",
                    "penghasilan_ayah",
                    "sampingan_ayah",
                    "scan_ktp_ayah",
                    "scan_slip_ayah",
                    "tempat_lahir_ayah",
                    "tanggal_lahir_ayah",
                    "atribut",
                    "executor",
                    "timestamp",
                ],
            }
        )
            .then(function (response) {
                resolve(response);
            })
            .catch(function (err) {
                reject(err);
            });
    });
};

ayah.prototype.getByLoggedIn = function (no_peserta) {
    return new Promise(function (resolve, reject) {
        Ayah.findOne({
            where: {
                no_peserta: no_peserta,
            },
            include: [Provinsi, Kabkot, Kecamatan, Pekerjaan],
        })
            .then(function (response) {
                resolve(response);
            })
            .catch(function (err) {
                reject(err);
            });
    });
};

ayah.prototype.checkData = function (no_peserta, users) {
    return new Promise((resolve, reject) => {
        Ayah.findOne({
            where: {
                no_peserta: no_peserta,
            },
            raw: true,
        })
            .then(function (response) {
                var result = true;
                var data = {};
                if (response.status_ayah === "wafat") {
                    data.nama_ayah = response.nama_ayah;
                } else {
                    if (users.ukt_tinggi === "ya") {
                        delete response.nik_ayah;
                        delete response.pekerjaan_ayah;
                        delete response.penghasilan_ayah;
                        delete response.scan_ktp_ayah;
                        delete response.scan_slip_ayah;
                    }

                    // Jika tidak bekerja
                    if (response.pekerjaan_ayah === "11") {
                        delete response.penghasilan_ayah;
                        delete response.scan_slip_ayah;
                    }

                    delete response.sampingan_ayah;
                    // delete response.telepon_ayah
                    data = response;
                }
                for (var key in data) {
                    result *= !!data[key];
                    console.log({ key, value: data[key] });
                    // result *= data[key] !== null && data[key] !== "" && data[key] !== 0;
                }
                resolve(result);
            })
            .catch(function (err) {
                reject(err);
            });
    });
};

module.exports = ayah;
